### ISSUE TITLE
Ship::JumpsRemaining, Wormholes, and AI::IsStranded()

### ISSUE THREAD

**Issue by tehhowch** [endless-sky/endless-sky#2829]
There is some issue on `master` which causes a ship to retain a `targetSystem` despite being in that system. This results in JumpsRemaining() returning 0, and AI::IsStranded() to return `true`, even if that ship has full fuel reserves and is otherwise very capable of jumping or landing on the wormhole. I think I have observed this for both parents and escorts (e.g. as a result of both `MoveEscort()` and `MoveIndependent()`, but am at minimum certain it can occur for parents.

This combination results in streams of assisting ships that all try to refuel a ship that has plenty of fuel.
![assistance swarm](https://user-images.githubusercontent.com/20871346/28636182-f591184e-7202-11e7-8817-20d42985986d.png)
This ship is parentless, has no escorts, has a target system of Sabriset (we're in Sabriset), and has the idea to land on the wormhole, but cannot ever get there since it keeps waiting to be refueled (despite having full fuel). If I use the debugger to set its `targetSystem` to 0, then it is magically free to move about.

I've tried to track down the issue, as it seems to be correlated with a wormhole being present in the system (e.g. a landable planet, but no planets with fuel), but wormhole travel routing (in MoveEscort) always sets `targetSystem` to be `nullptr` before assigning the `targetPlanet`, and MoveIndependent seems incapable of setting the wormhole object as a landing destination unless the ship is actually out of fuel (since the wormhole has no spaceport, and [planets with no spaceport are not considered](https://github.com/endless-sky/endless-sky/blob/fa92d76902aa6f63dae1187c001851c746723a85/source/AI.cpp#L1016) until jumping is not a possibility).

The effect of this erroneous `JumpsRemaining() = 0` could be hidden by adding a check during the "can anyone help me" phase of [AI::Step#351](https://github.com/endless-sky/endless-sky/blob/fa92d76902aa6f63dae1187c001851c746723a85/source/AI.cpp#L351), to only accept help if the other ship can refuel it and it needs refueling, e.g.
```
if(isStranded && !ship->CanRefuel(*it))
	continue;
```
becomes
```
if(isStranded && !ship->CanRefuel(*it) && it->JumpFuelMissing())
	continue;
```
since [JumpFuelMissing()](https://github.com/endless-sky/endless-sky/blob/fa92d76902aa6f63dae1187c001851c746723a85/source/Ship.cpp#L1979) will return 0 if the ship's `targetSystem` is the same as its current system (e.g. it's not a valid destination for `JumpFuel()`, so it needs 0 fuel to reach).

But, hiding the effect doesn't solve the underlying issue, or resolve the cause of the issue.

Another possibility is to forcibly set `targetSystem = nullptr` when doing wormhole travel in [Ship::Move](https://github.com/endless-sky/endless-sky/blob/fa92d76902aa6f63dae1187c001851c746723a85/source/Ship.cpp#L1030), but this may affect the travel routing that happens in AI.cpp (I have not deeply investigated this), and does not address the cause of the issue.

**TLDR**: Some AI ships think they are stranded when they are not, because they have an erroneous `targetSystem` value, and I'd appreciate help tracking down the reason for the erroneous value.

**Comment by tehhowch** [comment]
An easy way to reproduce, from Steam: https://youtu.be/CIWjTBWTJec
The user's save: 
[Travion Expectra.txt](https://github.com/endless-sky/endless-sky/files/1180193/Travion.Expectra.txt)

Jump to Segesta
Wait for fleet to arrive
Set a travel plan to go to Cardea
Invoke autopilot via either J or L
When arriving in Cardea, hold J. As the fleet arrives, they will all indicate that that they do not have fuel to jump. However, if you actually do jump to Antevorta, they will follow.

**Comment by tehhowch** [comment]
git bisect testing indicates that the AI never thought it was stranded when exiting wormholes before c795aae (e.g. when JumpsRemaining began calling `JumpFuel(targetSystem)`), but will think some ships (sometimes all, but not always) are in need of fuel after passing through a wormhole. As soon as their parent leaves the system (if they have a parent), they will correctly route to the parent (because targetSystem has now been set to an appropriate value, so they know they have enough fuel to jump). As mentioned above, this is the symptom of the underlying issue - e.g. the real issue existed before c795aae.

**Comment by tehhowch** [comment]
Alright, found the issue (at minimum, a very related issue). It's an old one, too! And 100% related to wormhole travel and escort routing. The line at fault is [AI.cpp#L1135: ship.SetTargetSystem(to)](https://github.com/tehhowch/endless-sky/blob/master/source/AI.cpp#L1135)

Commit 3225ee59c8 introduced wormhole pathfinding, but if we look at this commit, we see that this faulty line is appropriately guarded, and only called when the route between `from` and `to` has a cost (e.g., is a jump of some kind). If it's a wormhole landing, then there is no setting of a target system.
Commit b51c57efd4 updated that check, but kept it intact
Commit a1347562239 accidentally did away with this blocking when moving the DistanceMap route calculation - the `break` called when finding a wormhole that leads to the desired system doesn't escape the SetTargetSystem, so the ship gets both a targetStellar to land on, and a targetSystem to jump to. Thus when c795aae was published and JumpsRemaining considered `targetSystem`, escorts which went through a wormhole could start thinking they were stranded.

I've got a fix(#2839) that actually is a two-fer: it will solve the issue with escorts trying to refuel themselves after going through a wormhole, and also fix the "landing dance" that happens if an escort has no route to the system you are jumping to. That landing dance is a very old bug, and originated in 27c16e1778f2f :  If `dest` doesn't exist (e.g. unroutable), then `(!dest || ....)` returns true, and the ship is told to go refuel. So if a ship has full fuel, and you are preparing to jump someplace it can't follow, it would continuously be told to refuel. With the fix, it will instead fall through to KeepStation with your ship, until you jump away (and only then will it refuel, and then wait in the system center).

**Comment by tehhowch** [comment]
This can be closed with the milestone 👍
